Fix #2972, ensure masked_where gives structured mask for structured masked array

gerritholl opened the pull request:
This pull request fixes bug #2972, caused by a bug in `numpy.ma.masked_where` when it is passed a
structured array.  `masked_where` was inadvertently setting the _mask field
of a structured array to a non-structured array, so we would end up with
a masked structured array where `x.data` was structured, but `x.mask` was not.
This would lead to troubles for methods such as `__getitem__`,`**repr**`, and
`tolist`.  By writing to`.mask`instead of`._mask`, this problem is
mitigated.

This is an alternative solution to pull request #5513.

Commit:
BUG: Fix for issue #2972

This commit fixes a bug in numpy.ma.masked_where when it is passed a
structured array.  masked_where was inadvertently setting the _mask field
of a structured array to a non-structured array, so we would end up with
a masked structured array where x.data was structured, but x.mask was not.
This would lead to troubles for methods such as __getitem__, __repr__, and
tolist.  By writinge to .mask instead of ._mask, this problem is
mitigated.

Commit:
TST: Add test for fix to issue #2972

Add a test to TestMaskedArrayFunctions to verify that a masked array
created with masked_where has a structured mask.  This was previously
unstructured, leading to bug #2972.  Commit for original bugfix was
ca4e201fbd0f0a79759d07b357a5a930e4c79dd8.

jaimefrio commented:
@hpaulj you seem to have given this issue quite some thought, do you have an opinion on the approaches of these 2 related PRs?

@gerritholl You should add the exact same failures you reported in #2972 as a test, it is the best way to be sure that the issue is fixed if the implementation is changed.

gerritholl commented:
@jaimefrio I'm confused, isn't this the same already?  The only change compared to what I did in #2972 is the variable name, and that I call `np.range` rather than `np.zeros`.  That doesn't seem to matter for the test.  Or are you referring to something else?

jaimefrio commented:
Well, in #2972 the call `am = numpy.ma.masked_where(a["A"]<5, a)` doesn't fail, what fails is when you afterwards try to access `am['A']`.

In the test here you are checking that `assert_equal(am.mask.dtype.names, am.dtype.names)`, which is your assessment of what is causing the failure with the current implementation. I think the test should check also that calling `am['A']` does not fail.

gerritholl commented:
Oh, right.  I see.  I'll revise the test tomorrow.

rgommers commented on numpy/ma/tests/test_core.py:
```diff
@@ -2985,6 +2985,13 @@ def test_masked_where_shape_constraint(self):
+        # test that masked_where on a structured array sets a structured
+        # mask (see issue #2972)
+        a = np.zeros(10, dtype=[("A", "<f2"), ("B", "<f4")])
+        am = numpy.ma.masked_where(a["A"]<5, a)
```
``numpy`` --> ``np``

rgommers commented:
@gerritholl can you update the test?

Commit:
TST: Fix test for fix to issue #2972

Expand test for fix for issue #2972 to be more thorough.  Commit for
original bugfix was ca4e201fbd0f0a79759d07b357a5a930e4c79dd8 and commit
for original test was d764f02ebe011e83a540c65abc7b6c1e6055425d.

charris commented on numpy/ma/core.py:
```diff
@@ -1815,7 +1815,7 @@ def masked_where(condition, a, copy=True):
         cls = MaskedArray
     result = a.view(cls)
-    result._mask = cond
+    result.mask = cond
```
Could you add a comment here that we need to use the property `.mask` rather than `._mask` to assure that the mask is set correctly for structured arrays?

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -2985,6 +2985,15 @@ def test_masked_where_shape_constraint(self):
+        # test that masked_where on a structured array sets a structured
+        # mask (see issue #2972)
+        a = np.zeros(10, dtype=[("A", "<f2"), ("B", "<f4")])
+        am = np.ma.masked_where(a["A"]<5, a)
```
Spaces around `<`.

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -2985,6 +2985,15 @@ def test_masked_where_shape_constraint(self):
+        am = np.ma.masked_where(a["A"]<5, a)
+        assert_equal(am.mask.dtype.names, am.dtype.names)
+        assert_equal(am["A"],
+                    np.ma.masked_array(np.zeros(10), np.ones(10)))
```
Needs extra space for alignment.

charris commented:
Would it make sense to also include the test from #5513 (which needs cleanup for style).

charris commented:
Is the fix in #5513 orthogonal to this, or is it never needed? As a fix for #2972 I prefer this, but want to make sure that there is nothing in #5513 that also needs inclusion before I close it.

charris commented:
`mask_invalid` may also need this fix. It says it calls `masked_where`, but it doesn't.

charris commented:
Cleaned up a bit in #5824, so closing this. Thanks @gerritholl.
